src/pthread-main.js

```diff
@@ -52,19 +52,39 @@ Module['print'] = threadPrint;
 Module['printErr'] = threadPrintErr;
 this.alert = threadAlert;
 
+Module['instantiateWasm'] = function(info, receiveInstance) {
+  // Instantiate from the module posted from the main thread.
+  // We can just use sync instantiation in the worker.
+  instance = new WebAssembly.Instance(Module['wasmModule'], info);
+  // We don't need the module anymore; new threads will be spawned from the main thread.
+  // XXX is this true?
+  Module['wasmModule'] = undefined;
+  receiveInstance(instance);
+  return instance.exports;
+}
+
 this.onmessage = function(e) {
   try {
     if (e.data.cmd === 'load') { // Preload command that is called once per worker to parse and load the Emscripten code.
       // Initialize the thread-local field(s):
       tempDoublePtr = e.data.tempDoublePtr;
 
       // Initialize the global "process"-wide fields:
-      buffer = e.data.buffer;
       Module['TOTAL_MEMORY'] = TOTAL_MEMORY = e.data.TOTAL_MEMORY;
       STATICTOP = e.data.STATICTOP;
       DYNAMIC_BASE = e.data.DYNAMIC_BASE;
       DYNAMICTOP_PTR = e.data.DYNAMICTOP_PTR;
 
+
+      if (e.data.wasmModule) {
```

Comment:
Let's make this a `#if WASM` as well, because in Wasm mode, `wasmModule` is required to exist - if it doesn't we are in trouble, so in `#if WASM` we cannot proceed if `e.data.wasmModule` did not exist.